src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets

```diff
@@ -162,18 +162,57 @@ Copyright (c) .NET Foundation. All rights reserved.
     
+    <_FrameworkVersionForImplicitDefine>$(_FrameworkVersionForImplicitDefine.Replace('.', '_'))</_FrameworkVersionForImplicitDefine>
+
     <_FrameworkVersionForImplicitDefine Condition=" '$(TargetFrameworkIdentifier)' == '.NETFramework'">$(_FrameworkVersionForImplicitDefine.Replace('_', ''))</_FrameworkVersionForImplicitDefine>
 
     <ImplicitFrameworkDefine>$(_FrameworkIdentifierForImplicitDefine)$(_FrameworkVersionForImplicitDefine)</ImplicitFrameworkDefine>
   </PropertyGroup>
 
+  <!-- Add conditional compilation symbols for the target platform (for example ANDROID, IOS, WINDOWS) -->
+  <Target Name="GenerateTargetPlatformDefineConstants"
+        Condition=" '$(DisableImplicitFrameworkDefines)' != 'true' and '$(TargetPlatformIdentifier)' != '' and '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionLessThanOrEquals(5.0, $(TargetFrameworkVersion))) "
+        BeforeTargets="Build" >
+    <PropertyGroup>
+      <_PlatformIdentifierForImplicitDefine>$(TargetPlatformIdentifier.ToUpperInvariant())</_PlatformIdentifierForImplicitDefine>
+      <_PlatformVersionForImplicitDefine>$(TargetPlatformVersion.Replace('.', '_'))</_PlatformVersionForImplicitDefine>
+    </PropertyGroup>
+    <ItemGroup>
+      <_ImplicitDefineConstant Include="$(_PlatformIdentifierForImplicitDefine)" />
+      <_ImplicitDefineConstant Include="$(_PlatformIdentifierForImplicitDefine)$(_PlatformVersionForImplicitDefine)" />
+    </ItemGroup>
+  </Target>
+  
+  <!-- Add conditional compilation symbols for target framework backwards compatibility with .NET 5.0 and later (for example .NET 6.0 defines NET5_0 and NETCOREAPP3_1) -->
+  <Target Name="GenerateNETCompatibleDefineConstants"
+          Condition=" '$(DisableImplicitFrameworkDefines)' != 'true' and '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionLessThanOrEquals(5.0, $(TargetFrameworkVersion))) "
+          BeforeTargets="Build" >
+    <ItemGroup>
+      <_NetCoreAppFrameworkVersions Include="@(SupportedNETCoreAppTargetFramework->'%(Identity)'->TrimStart('.NETCoreApp,Version=v'))" />
+      <_NetCoreAppCompatibleVersions Include="@(_NetCoreAppFrameworkVersions)" Condition=" $([MSBuild]::VersionGreaterThan(%(Identity), 3.1)) and $([MSBuild]::VersionLessThan(%(Identity), $(TargetFrameworkVersion))) " />
+      <_ImplicitDefineConstant Include="NETCOREAPP" />
+      <_ImplicitDefineConstant Include="NETCOREAPP3_1" />
+      <_ImplicitDefineConstant Include="@(_NetCoreAppCompatibleVersions->'NET%(Identity)'->Replace('.', '_'))" />
+    </ItemGroup>
+  </Target>
+
+  <!-- Add implicitly defined preprocessor symbols to DefineConstants -->
+  <Target Name="AddImplicitDefineConstants"
+          Condition=" '$(DisableImplicitFrameworkDefines)' != 'true' "
+          DependsOnTargets="GenerateTargetPlatformDefineConstants;GenerateNETCompatibleDefineConstants"
+          BeforeTargets="Build" >
```

Comment:
I've updated to overwrite `CoreCompile` with the target that writes `DefineConstants` to our output file. I overwrote compilation instead of running the target afterwards because the net6.0 tests won't compile yet so the build fails before I can get output. If there's a better way to do it, let me know.